GameEngine/Agents/BeamStackAgent.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            var solution = BeamStack(new RootNode(state), int.MaxValue).ToList();
+
+            if (solution.Count == 0)
+            {
+                throw new NoMoveFoundException();
+            }
+
+            return solution.First().Action;
+        }
+
+        private IEnumerable<SearchNode> BeamStack(SearchNode start, int upperBound)
+        {
+            var beamStack = new Stack<MinMax>();
+            beamStack.Push(new MinMax { Min = 0, Max = upperBound });
+
+            SearchNode optimalSolution = null;
+            while(beamStack.Count() != 0)
+            {
+                var solution = Search(beamStack, start, upperBound);
+                if (solution != null)
+                {
+                    optimalSolution = solution;
+                    upperBound = _heuristic.Evaluate(solution.State);
+                }
+                while(beamStack.Count() != 0 && beamStack.Peek().Max >= upperBound)
+                {
+                    beamStack.Pop();
+                }
+                if(beamStack.Count() != 0)
+                {
+                    beamStack.Peek().Min = beamStack.Peek().Max;
+                    beamStack.Peek().Max = upperBound;
+                }
+            }
+
+            if (optimalSolution == null)
+            {
+                throw new NoMoveFoundException();
+            }
+
+            return optimalSolution.Solution();
+        }
+
+        private SearchNode Search(Stack<MinMax> beamStack, SearchNode start, int upperBound)
+        {
+            var currentLayer = new List<SearchNode> { start };
+            var nextLayer = new List<SearchNode>();
+
+            SearchNode bestGoal = null;
+
+            while(currentLayer.Count != 0 || nextLayer.Count != 0)
+            {
+                while(currentLayer.Count != 0)
+                {
+                    var node = FindCheapest(currentLayer);
+                    currentLayer.Remove(node);
+                    if (node.State.IsOver)
+                    {
+                        upperBound = _heuristic.Evaluate(node.State);
+                        bestGoal = node;
+                    }
+                    else
+                    {
+                        nextLayer.AddRange(GetSuccessorsInRange(node, beamStack.Peek()));
+                        if (nextLayer.Count > _width)
+                        {
+                            nextLayer = Prune(nextLayer, beamStack.Peek());
+                        }
+                    }
+                }
+                currentLayer = nextLayer;
+                nextLayer = new List<SearchNode>();
+                beamStack.Push(new MinMax { Min = 0, Max = upperBound });
+            }
+
+            return bestGoal;
+        }
+        
+        private List<SearchNode> Prune(List<SearchNode> layer, MinMax range)
+        {
+            var highestCost = int.MinValue;
+            var nodes = new List<Tuple<SearchNode, int>>();
```

Comment:
💡 I'm not a fan of Tuples just because `Item1` and `Item2` are so... contextless.

I don't think there's an easy way to use anonymous types here. What do you think about declaring a private class/struct within this file?

This doesn't hold up merging.